extensions-contrib/virtual-columns/src/main/java/org/apache/druid/segment/StringTypeMapVirtualColumnDimensionSelector.java

```diff
@@ -0,0 +1,210 @@
+    } else {
+      return new SingleIndexedInt(valueIndex);
+    }
+  }
+
+  @Override
+  public ValueMatcher makeValueMatcher(@Nullable String value)
+  {
+    return new ValueMatcher()
+    {
+      @Override
+      public boolean matches()
+      {
+        return Objects.equals(value, getObject());
+      }
+
+      @Override
+      public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+      {
+        inspector.visit("keySelector", getKeySelector());
+        inspector.visit("valueSelector", getValueSelector());
+        inspector.visit("subColumnName", subColumnName);
+      }
+    };
+  }
+
+  @Override
+  public ValueMatcher makeValueMatcher(Predicate<String> predicate)
+  {
+    return new ValueMatcher()
+    {
+      @Override
+      public boolean matches()
+      {
+        return predicate.apply((String) getObject());
+      }
+
+      @Override
+      public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+      {
+        inspector.visit("keySelector", getKeySelector());
+        inspector.visit("valueSelector", getValueSelector());
+        inspector.visit("subColumnName", subColumnName);
+      }
+    };
+  }
+
+  @Override
+  public int getValueCardinality()
+  {
+    // To get the value cardinarlity, we need to first check all keys and values to find valid pairs, and then find the
+    // number of distinct values among them.
+    return CARDINALITY_UNKNOWN;
+  }
+
+  @Nullable
+  @Override
+  public String lookupName(int id)
+  {
+    final DimensionSelector keySelector = getKeySelector();
+    final DimensionSelector valueSelector = getValueSelector();
+
+    final IndexedInts keyIndices = keySelector.getRow();
+    final IndexedInts valueIndices = valueSelector.getRow();
+
+    final int limit = Math.min(keyIndices.size(), valueIndices.size());
+    final int valueIndex = IntStream
+        .range(0, limit)
+        .filter(i -> subColumnName.equals(keySelector.lookupName(keyIndices.get(i)))) // subColumnName is never null
+        .findAny()
+        .orElse(-1);
+
+    if (valueIndex == id) {
+      return valueSelector.lookupName(id);
+    } else {
+      return null;
+    }
+  }
+
+  @Override
+  public boolean nameLookupPossibleInAdvance()
+  {
+    return false;
+  }
+
+  @Nullable
+  @Override
+  public IdLookup idLookup()
+  {
+    final DimensionSelector keySelector = getKeySelector();
+    final DimensionSelector valueSelector = getValueSelector();
+
+    final IdLookup valueLookup = valueSelector.idLookup();
+
+    if (valueLookup != null) {
+      final IndexedInts keyIndices = keySelector.getRow();
+      final IndexedInts valueIndices = valueSelector.getRow();
+
+      final int limit = Math.min(keyIndices.size(), valueIndices.size());
+      final int valueIndex = IntStream
+          .range(0, limit)
+          .filter(i -> subColumnName.equals(keySelector.lookupName(keyIndices.get(i)))) // subColumnName is never null
+          .findAny()
+          .orElse(-1);
```

Comment:
this code is partly repeated many times in this class, can we extract a common method ?

Reply:
Done.